[charts] Add control to the axis highlight (#17900)
mui/mui-x: docs/data/charts/highlighting/highlighting.md

{{"demo": "ElementHighlights.js"}}

## Controlled item highlight

The item highlight can be controlled by using `highlightedItem` and `onHighlightChange`.

You can set the `highlightedItem` value based on inputs, and sync it when the user hover over an item themselves.

{{"demo": "ControlledHighlight.js"}}

## Controlled axis highlight

The highlight can be controlled by using `highlightedAxis` prop.

> Its value can be `null` to remove axis highlight, or an object `{ direction: 'x' | 'y', axisId: string, dataIndex: number }`.

Reviewer: ```suggestion
Its value can be `null` to remove axis highlight, or an object `{ axisId: string, dataIndex: number }`.
```

> The `onAxisInteraction` handler is trigger each time the pointer crosses the boundaries between two axis values.

Reviewer: ```suggestion
The `onAxisInteraction` handler is triggered each time the pointer crosses the boundaries between two axis values.
```

Its parameter is an array of objects `{ direction, axisId, dataIndex }`.

> One per axis.

Reviewer: So, if no axis highlight exists, the argument is an empty array, right?

> One per axis.

How does this work? I'm trying the example below and there's an x and a y axes, but `onAxisInteraction` only returns: 

```json
[
    {
        "axisId": "x-axis",
        "dataIndex": 4
    }
]
```

Which axes are supposed to be shown?

Axes without data are ignored by the handler.

> The handler gets an array of objects.

Reviewer: Isn't this the same as the content on line 60?

Whereas the controlled value only accept one object.

:::warning

> The `onAxisInteraction` can be triggered at a high frequency when user move their pointer on the chart.

Reviewer: ```suggestion
The `onAxisInteraction` can be triggered at a high frequency when the user moves their pointer over the chart.
```

To avoid performance issues, avoid defining object in the props which creates a new reference at each render.
Either define them outside the component, or memoize them.

> Especially for axes and series which when updated impact a lot of computation.

Reviewer: ```suggestion
This suggestion is especially useful for axes and series which, when updated, impact a lot of computation.
```

```jsx

> // ❌ The chart gets a new axis at each render. Leading to useless computation.

Reviewer: ```suggestion
// ❌ The chart gets a new axis on each render, leading to useless computation.
```

## Controlled highlight

### Item

The item highlight can be controlled by using `highlightedItem` and `onHighlightChange`.

You can set the `highlightedItem` value based on inputs, and sync it when the user hover over an item themselves.

{{"demo": "ControlledHighlight.js"}}

### Axis

The highlight can be controlled by using `xAxisHighlight`/`yAxisHighlight` and the `onXAxisInteraction`/`onYAxisInteraction`.

The `xAxisHighlight`/`yAxisHighlight` are objects `{ axisId, dataIndex, value }`.

> If the `dataIndex` is provided, the axis `value` is ignored an computed from the `dataIndex` and `axis.data`.

Reviewer 2: What do you mean with 

> If the `dataIndex` is provided

provided by whom?

> The `onXAxisInteraction`/`onYAxisInteraction` handler are trigger each time pointer moves from one axis value to another.

Reviewer 2: ```suggestion
The `onXAxisInteraction`/`onYAxisInteraction` handler are triggered each time the pointer moves from one axis value to another.
```

> The `onXAxisInteraction`/`onYAxisInteraction` handler are trigger each time pointer moves from one axis value to another.

Reviewer 2: Maybe we want to rewrite most text instead 

```suggestion
The `onXAxisInteraction`/`onYAxisInteraction` handler are triggered each time the pointer crosses the boundaries between two axes.
```

Its parameter is an array of objects `{ axisId, dataIndex, value }`.
One per axis.

:::warning

> The handler get an array of axis value identifier.
> Whereas the controlled value only accept on objects.

Reviewer 2: ```suggestion
The handler accepts an array of axis value identifiers.
Whereas the controlled value only accepts one object.
```

For now highlight components assume you use the first axis.

> Being able to highlight any axis will arrive in further development.

Reviewer 2: doesnt seem to add much from the phrase above it

```suggestion
```
